feat: Add thinking and reasoning_effort parameter support for GitHub Copilot provider

timelfrink opened the pull request:
- Add github_copilot case to get_supported_openai_params function
- Implement get_supported_openai_params method in GithubCopilotConfig
- Dynamically add thinking and reasoning_effort params for Anthropic models
- Add comprehensive tests for parameter support validation
- Ensure case-insensitive model detection for parameter inclusion

Fixes UnsupportedParamsError when using advanced reasoning parameters with Anthropic models through GitHub Copilot proxy.

## Title

Add thinking and reasoning_effort parameter support for GitHub Copilot provider

## Relevant issues
``` code
If you want to use these params dynamically send allowed_openai_params=['thinking'] in your request.
Traceback (most recent call last):
  File "/Users/elfrit01/Documents/Stepstone/copilot-litellm-proxy/.venv/lib/python3.11/site-packages/litellm/llms/anthropic/experimental_pass_through/adapters/handler.py", line 157, in async_anthropic_messages_handler
    completion_response = await litellm.acompletion(**completion_kwargs)
                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/elfrit01/Documents/Stepstone/copilot-litellm-proxy/.venv/lib/python3.11/site-packages/litellm/utils.py", line 1586, in wrapper_async
    raise e
  File "/Users/elfrit01/Documents/Stepstone/copilot-litellm-proxy/.venv/lib/python3.11/site-packages/litellm/utils.py", line 1437, in wrapper_async
    result = await original_function(*args, **kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/elfrit01/Documents/Stepstone/copilot-litellm-proxy/.venv/lib/python3.11/site-packages/litellm/main.py", line 563, in acompletion
    raise exception_type(
  File "/Users/elfrit01/Documents/Stepstone/copilot-litellm-proxy/.venv/lib/python3.11/site-packages/litellm/main.py", line 536, in acompletion
    init_response = await loop.run_in_executor(None, func_with_context)
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/elfrit01/.pyenv/versions/3.11.0/lib/python3.11/concurrent/futures/thread.py", line 58, in run
    result = self.fn(*self.args, **self.kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/elfrit01/Documents/Stepstone/copilot-litellm-proxy/.venv/lib/python3.11/site-packages/litellm/utils.py", line 1060, in wrapper
    result = original_function(*args, **kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/elfrit01/Documents/Stepstone/copilot-litellm-proxy/.venv/lib/python3.11/site-packages/litellm/main.py", line 3525, in completion
    raise exception_type(
  File "/Users/elfrit01/Documents/Stepstone/copilot-litellm-proxy/.venv/lib/python3.11/site-packages/litellm/main.py", line 1246, in completion
    optional_params = get_optional_params(
                      ^^^^^^^^^^^^^^^^^^^^
  File "/Users/elfrit01/Documents/Stepstone/copilot-litellm-proxy/.venv/lib/python3.11/site-packages/litellm/utils.py", line 3357, in get_optional_params
    _check_valid_arg(
  File "/Users/elfrit01/Documents/Stepstone/copilot-litellm-proxy/.venv/lib/python3.11/site-packages/litellm/utils.py", line 3340, in _check_valid_arg
    raise UnsupportedParamsError(
litellm.exceptions.UnsupportedParamsError: litellm.UnsupportedParamsError: github_copilot does not support parameters: ['thinking'], for model=claude-sonnet-4. To drop these, set `litellm.drop_params=True` or for proxy:

`litellm_settings:
 drop_params: true`
```
## Pre-Submission checklist

**Please complete all items before asking a LiteLLM maintainer to review your PR**



- [x] I have Added testing in the [`tests/litellm/`](https://github.com/BerriAI/litellm/tree/main/tests/litellm) directory, **Adding at least 1 test is a hard requirement** - [see details](https://docs.litellm.ai/docs/extras/contributing_code)
- [x] I have added a screenshot of my new test passing locally 
- [x] My PR passes all unit tests on [`make test-unit`](https://docs.litellm.ai/docs/extras/contributing_code)
- [x] My PR's scope is as isolated as possible, it only solves 1 specific problem


## Type
🆕 New Feature

## Changes
Added support for thinking in github copilot proxy for Anthropic models.

<img width="884" height="146" alt="image" src="https://github.com/user-attachments/assets/50d7c846-f54d-4316-96b6-1a4c7f952d7e" />

timelfrink committed:
feat: Add thinking and reasoning_effort parameter support for GitHub Copilot provider

- Add github_copilot case to get_supported_openai_params function
- Implement get_supported_openai_params method in GithubCopilotConfig
- Dynamically add thinking and reasoning_effort params for Anthropic models
- Add comprehensive tests for parameter support validation
- Ensure case-insensitive model detection for parameter inclusion

Fixes UnsupportedParamsError when using advanced reasoning parameters
with Anthropic models through GitHub Copilot proxy.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/BerriAI/litellm?pullRequest=13691) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/BerriAI/litellm?pullRequest=13691) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/BerriAI/litellm?pullRequest=13691) it.</sub>

timelfrink committed:
formatting

timelfrink committed:
feat: add thinking and reasoning parameter support for GitHub Copilot provider

- Add dynamic parameter support for anthropic models through GitHub Copilot
- Include thinking parameter for anthropic model compatibility
- Support reasoning_effort parameter for both anthropic and reasoning models
- Update test coverage for parameter validation logic
- Ensure proper parameter filtering based on model type

ishaan-jaff reviewed (changes requested):
-

timelfrink committed:
fix: restrict thinking/reasoning_effort parameters to models with extended thinking support

Only models in the 4 family and 3-7 family support extended thinking features.
Previously all models would incorrectly receive these parameters.

Now uses supports_reasoning() to check model registry for actual capability.

timelfrink committed:
fix: remove redundant github_copilot check in get_supported_openai_params

The provider_config_manager already handles github_copilot provider
through LlmProviders.GITHUB_COPILOT mapping, making the explicit
check unnecessary.

timelfrink commented on litellm/litellm_core_utils/get_supported_openai_params.py:
```diff
@@ -266,11 +266,12 @@ def get_supported_openai_params(  # noqa: PLR0915
+            return ElevenLabsAudioTranscriptionConfig().get_supported_openai_params(
+                model=model
             )
+    elif custom_llm_provider == "github_copilot":
```
Fixed - removed the redundant check since provider_config_manager already handles this.

timelfrink committed:
fix: resolve case sensitivity and test failures for extended thinking support

- Fix supports_reasoning() call to use lowercase model names for proper lookup
- Remove custom_llm_provider parameter as model registry entries are provider-agnostic
- Update tests to use full model names with date stamps (required for supports_reasoning)
- Add test coverage for models without extended thinking support

timelfrink committed:
Merge remote-tracking branch 'origin/main' into feat/github-copilot-thinking-reasoning-support

timelfrink committed:
Merge branch 'main' of https://github.com/BerriAI/litellm into feat/github-copilot-thinking-reasoning-support

Commit:
Merge branch 'main' into feat/github-copilot-thinking-reasoning-support
